go/vt/vtgate/engine/fk_cascade.go

```diff
@@ -82,77 +97,149 @@ func (fkc *FkCascade) TryExecute(ctx context.Context, vcursor VCursor, bindVars
+	bindVars[child.BVName] = bv
+	var err error
+	if isStreaming {
+		err = vcursor.StreamExecutePrimitive(ctx, child.Exec, bindVars, wantfields, func(result *sqltypes.Result) error { return nil })
+	} else {
+		_, err = vcursor.ExecutePrimitive(ctx, child.Exec, bindVars, wantfields)
+	}
+	if err != nil {
+		return err
+	}
+	delete(bindVars, child.BVName)
+	return nil
+}
+
+func (fkc *FkCascade) executeNonLiteralExprFkChild(ctx context.Context, vcursor VCursor, bindVars map[string]*querypb.BindVariable, wantfields bool, selectionRes *sqltypes.Result, child *FkChild, isStreaming bool) error {
+	// For each row in the SELECT we need to run the child primitive.
+	for _, row := range selectionRes.Rows {
+		// First we check if any of the columns is being updated at all.
+		skipRow := true
+		for _, info := range child.NonLiteralInfo {
+			// We use a null-safe comparison, so the value is guaranteed to be not null.
+			// We check if the column has updated or not.
+			isUnchanged, err := row[info.CompExprCol].ToBool()
+			if err != nil {
+				return err
+			}
+			if !isUnchanged {
+				// If any column has changed, then we can't skip this row.
+				// We need to execute the child primitive.
+				skipRow = false
+				break
+			}
+		}
+		// If none of the columns have changed, then there is no update to cascade, we can move on.
+		if skipRow {
+			continue
+		}
+		// We create a bindVariable that stores the tuple of columns involved in the fk constraint.
 		bv := &querypb.BindVariable{
 			Type: querypb.Type_TUPLE,
 		}
-		for _, row := range selectionRes.Rows {
-			var tupleValues []sqltypes.Value
-			for _, colIdx := range child.Cols {
-				tupleValues = append(tupleValues, row[colIdx])
-			}
-			bv.Values = append(bv.Values, sqltypes.TupleToProto(tupleValues))
+		// Create a tuple from the Row.
+		var tupleValues []sqltypes.Value
+		for _, colIdx := range child.Cols {
+			tupleValues = append(tupleValues, row[colIdx])
 		}
+		bv.Values = append(bv.Values, sqltypes.TupleToProto(tupleValues))
 		// Execute the child primitive, and bail out incase of failure.
 		// Since this Primitive is always executed in a transaction, the changes should
-		// be rolled back incase of an error.
+		// be rolled back in case of an error.
 		bindVars[child.BVName] = bv
-		_, err = vcursor.ExecutePrimitive(ctx, child.Exec, bindVars, wantfields)
+
+		// Next, we need to copy the updated expressions value into the bind variables map.
+		for _, info := range child.NonLiteralInfo {
+			bindVars[info.UpdateExprBvName] = sqltypes.ValueBindVariable(row[info.UpdateExprCol])
+		}
+		var err error
+		if isStreaming {
+			err = vcursor.StreamExecutePrimitive(ctx, child.Exec, bindVars, wantfields, func(result *sqltypes.Result) error { return nil })
+		} else {
+			_, err = vcursor.ExecutePrimitive(ctx, child.Exec, bindVars, wantfields)
+		}
 		if err != nil {
-			return nil, err
+			return err
 		}
+		// Remove the bind variables that have been used and are no longer required.
```

Comment:
instead of doing this, don't we want to `map.Clone` at the beginning of this method and then just throw away the copy at the end?

Reply:
Done